
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Ag3 cloud data access &#8212; MalariaGEN vector data user guide</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet" />
  <link href="../_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.acff12b8f9c144ce68a297486a2fa670.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/myfile.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Ag3 data downloads" href="download.html" />
    <link rel="prev" title="Introduction to Ag3" href="intro.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">MalariaGEN vector data user guide</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../landing-page.html">
   MalariaGEN vector data user guide
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Ag3 (Ag1000G phase 3)
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="intro.html">
   Introduction to Ag3
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Ag3 cloud data access
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="download.html">
   Ag3 data downloads
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="examples.html">
   Ag3 analysis examples
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="examples/collection-sites.html">
     Mosquito collection sites
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="examples/snps.html">
     Single nucleotide polymorphisms (SNPs)
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="examples/pca.html">
     Investigating population structure with PCA
    </a>
   </li>
  </ul>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/ag3/cloud.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/malariagen/vector-data"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/malariagen/vector-data/issues/new?title=Issue%20on%20page%20%2Fag3/cloud.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://gke.mybinder.org/v2/gh/malariagen/vector-data/master?urlpath=tree/docs/ag3/cloud.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/malariagen/vector-data/blob/master/docs/ag3/cloud.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-hosting">
   Data hosting
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#sample-sets">
   Sample sets
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#sample-metadata">
   Sample metadata
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#snp-sites-and-alleles">
   SNP sites and alleles
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#site-filters">
   Site filters
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#snp-genotypes">
   SNP genotypes
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#snp-calls-dataset">
   SNP calls dataset
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#example-computation">
   Example computation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#running-larger-computations">
   Running larger computations
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#feedback-and-suggestions">
   Feedback and suggestions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#api-docs">
   API docs
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="ag3-cloud-data-access">
<h1>Ag3 cloud data access<a class="headerlink" href="#ag3-cloud-data-access" title="Permalink to this headline">¶</a></h1>
<p>This page provides information about how to access data from <a class="reference internal" href="intro.html"><span class="doc std std-doc"><em>Anopheles gambiae</em> 1000 Genomes project (Ag1000G) phase 3</span></a> via Google Cloud. This includes sample metadata and single nucleotide polymorphism (SNP) calls.</p>
<p>This notebook illustrates how to read data directly from the cloud, without having to first download any data locally. This notebook can be run from any computer, but will work best when run from a compute node within Google Cloud, because it will be physically closer to the data and so data transfer is faster. For example, this notebook can be run via <a class="reference external" href="https://gke.mybinder.org">MyBinder</a> or <a class="reference external" href="https://colab.research.google.com/">Google Colab</a> which are free interactive computing service running in the cloud.</p>
<p>To launch this notebook in the cloud and run it for yourself, click the launch icon (<i class="fas fa-rocket"></i>) at the top of the page and select one of the cloud computing services available.</p>
<div class="section" id="data-hosting">
<h2>Data hosting<a class="headerlink" href="#data-hosting" title="Permalink to this headline">¶</a></h2>
<p>All data required for this notebook is hosted on Google Cloud Storage (GCS). Data are hosted in the <code class="docutils literal notranslate"><span class="pre">vo_agam_release</span></code> bucket, which is a multi-region bucket located in the United States. All data hosted in GCS are publicly accessible and do not require any authentication to access.</p>
</div>
<div class="section" id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h2>
<p>Running this notebook requires some Python packages to be installed. These packages can be installed via pip or conda. E.g.:</p>
<div class="cell tag_hide-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>!pip install -q \
    zarr==2.6.1 \
    fsspec==0.8.7 \
    gcsfs==0.7.2 \
    dask==2021.03.0 \
    xarray==0.18.0 \
    scikit-allel==1.3.5 \
    malariagen_data==0.7.0
</pre></div>
</div>
</div>
</div>
<p>To make accessing these data more convenient, we’ve created the <a class="reference external" href="https://github.com/malariagen/malariagen-data-python">malariagen_data Python package</a>, which is available from PyPI. This is experimental so please let us know if you find any bugs or have any suggestions.</p>
<p>Now import the packages we’ll need to use here.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">dask</span>
<span class="kn">import</span> <span class="nn">dask.array</span> <span class="k">as</span> <span class="nn">da</span>
<span class="kn">from</span> <span class="nn">dask.diagnostics.progress</span> <span class="kn">import</span> <span class="n">ProgressBar</span>
<span class="c1"># silence some warnings</span>
<span class="n">dask</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="o">**</span><span class="p">{</span><span class="s1">&#39;array.slicing.split_large_chunks&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">})</span>
<span class="kn">import</span> <span class="nn">allel</span>
<span class="kn">import</span> <span class="nn">malariagen_data</span>
</pre></div>
</div>
</div>
</div>
<p>Ag3 data access from Google Cloud is set up with the following code:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span> <span class="o">=</span> <span class="n">malariagen_data</span><span class="o">.</span><span class="n">Ag3</span><span class="p">(</span><span class="s2">&quot;gs://vo_agam_release/&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="sample-sets">
<h2>Sample sets<a class="headerlink" href="#sample-sets" title="Permalink to this headline">¶</a></h2>
<p>Data in this release are organised into 28 sample sets. Each of these sample sets corresponds to a set of mosquito specimens contributed by a collaborating study. Depending on your objectives, you may want to access data from only specific sample sets, or all sample sets.</p>
<p>To see which sample sets are available, load the sample set manifest into a pandas dataframe:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_sample_sets</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">sample_sets</span><span class="p">()</span>
<span class="n">df_sample_sets</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample_set</th>
      <th>sample_count</th>
      <th>release</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AG1000G-AO</td>
      <td>81</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AG1000G-BF-A</td>
      <td>181</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AG1000G-BF-B</td>
      <td>102</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AG1000G-BF-C</td>
      <td>13</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AG1000G-CD</td>
      <td>76</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>5</th>
      <td>AG1000G-CF</td>
      <td>73</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>6</th>
      <td>AG1000G-CI</td>
      <td>80</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>7</th>
      <td>AG1000G-CM-A</td>
      <td>303</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>8</th>
      <td>AG1000G-CM-B</td>
      <td>97</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>9</th>
      <td>AG1000G-CM-C</td>
      <td>44</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>10</th>
      <td>AG1000G-FR</td>
      <td>23</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>11</th>
      <td>AG1000G-GA-A</td>
      <td>69</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>12</th>
      <td>AG1000G-GH</td>
      <td>100</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>13</th>
      <td>AG1000G-GM-A</td>
      <td>74</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>14</th>
      <td>AG1000G-GM-B</td>
      <td>31</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>15</th>
      <td>AG1000G-GM-C</td>
      <td>174</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>16</th>
      <td>AG1000G-GN-A</td>
      <td>45</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>17</th>
      <td>AG1000G-GN-B</td>
      <td>185</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>18</th>
      <td>AG1000G-GQ</td>
      <td>10</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>19</th>
      <td>AG1000G-GW</td>
      <td>101</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>20</th>
      <td>AG1000G-KE</td>
      <td>86</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>21</th>
      <td>AG1000G-ML-A</td>
      <td>60</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>22</th>
      <td>AG1000G-ML-B</td>
      <td>71</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>23</th>
      <td>AG1000G-MW</td>
      <td>41</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>24</th>
      <td>AG1000G-MZ</td>
      <td>74</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>25</th>
      <td>AG1000G-TZ</td>
      <td>300</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>26</th>
      <td>AG1000G-UG</td>
      <td>290</td>
      <td>v3</td>
    </tr>
    <tr>
      <th>27</th>
      <td>AG1000G-X</td>
      <td>297</td>
      <td>v3</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>For more information about these sample sets, see the section on sample sets in the <a class="reference internal" href="intro.html"><span class="doc std std-doc">introduction to Ag1000G phase 3</span></a>.</p>
</div>
<div class="section" id="sample-metadata">
<h2>Sample metadata<a class="headerlink" href="#sample-metadata" title="Permalink to this headline">¶</a></h2>
<p>Data about the samples that were sequenced to generate this data resource are available, including the time and place of collection, the gender of the specimen, and our call regarding the species of the specimen. These are organised by sample set.</p>
<p>E.g., load sample metadata for the AG1000G-BF-A and AG1000G-BF-B sample sets into a pandas dataframe:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_samples</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">(</span><span class="n">sample_sets</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;AG1000G-BF-A&quot;</span><span class="p">,</span> <span class="s2">&quot;AG1000G-BF-B&quot;</span><span class="p">])</span>
<span class="n">df_samples</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample_id</th>
      <th>partner_sample_id</th>
      <th>contributor</th>
      <th>country</th>
      <th>location</th>
      <th>year</th>
      <th>month</th>
      <th>latitude</th>
      <th>longitude</th>
      <th>sex_call</th>
      <th>sample_set</th>
      <th>release</th>
      <th>aim_fraction_colu</th>
      <th>aim_fraction_arab</th>
      <th>species_gambcolu_arabiensis</th>
      <th>species_gambiae_coluzzii</th>
      <th>species</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AB0085-Cx</td>
      <td>BF2-4</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Pala</td>
      <td>2012</td>
      <td>7</td>
      <td>11.150</td>
      <td>-4.235</td>
      <td>F</td>
      <td>AG1000G-BF-A</td>
      <td>v3</td>
      <td>0.024</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AB0086-Cx</td>
      <td>BF2-6</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Pala</td>
      <td>2012</td>
      <td>7</td>
      <td>11.150</td>
      <td>-4.235</td>
      <td>F</td>
      <td>AG1000G-BF-A</td>
      <td>v3</td>
      <td>0.038</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AB0087-C</td>
      <td>BF3-3</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Bana</td>
      <td>2012</td>
      <td>7</td>
      <td>11.233</td>
      <td>-4.472</td>
      <td>F</td>
      <td>AG1000G-BF-A</td>
      <td>v3</td>
      <td>0.982</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>coluzzii</td>
      <td>coluzzii</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AB0088-C</td>
      <td>BF3-5</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Bana</td>
      <td>2012</td>
      <td>7</td>
      <td>11.233</td>
      <td>-4.472</td>
      <td>F</td>
      <td>AG1000G-BF-A</td>
      <td>v3</td>
      <td>0.990</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>coluzzii</td>
      <td>coluzzii</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AB0089-Cx</td>
      <td>BF3-8</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Bana</td>
      <td>2012</td>
      <td>7</td>
      <td>11.233</td>
      <td>-4.472</td>
      <td>F</td>
      <td>AG1000G-BF-A</td>
      <td>v3</td>
      <td>0.975</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>coluzzii</td>
      <td>coluzzii</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>278</th>
      <td>AB0533-C</td>
      <td>BF13-18</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Souroukoudinga</td>
      <td>2014</td>
      <td>7</td>
      <td>11.235</td>
      <td>-4.535</td>
      <td>F</td>
      <td>AG1000G-BF-B</td>
      <td>v3</td>
      <td>0.021</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>279</th>
      <td>AB0536-C</td>
      <td>BF13-31</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Souroukoudinga</td>
      <td>2014</td>
      <td>7</td>
      <td>11.235</td>
      <td>-4.535</td>
      <td>F</td>
      <td>AG1000G-BF-B</td>
      <td>v3</td>
      <td>0.025</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>280</th>
      <td>AB0537-C</td>
      <td>BF13-32</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Souroukoudinga</td>
      <td>2014</td>
      <td>7</td>
      <td>11.235</td>
      <td>-4.535</td>
      <td>F</td>
      <td>AG1000G-BF-B</td>
      <td>v3</td>
      <td>0.029</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>281</th>
      <td>AB0538-C</td>
      <td>BF13-33</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Souroukoudinga</td>
      <td>2014</td>
      <td>7</td>
      <td>11.235</td>
      <td>-4.535</td>
      <td>F</td>
      <td>AG1000G-BF-B</td>
      <td>v3</td>
      <td>0.018</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>282</th>
      <td>AB0408-C</td>
      <td>BF14-20</td>
      <td>Austin Burt</td>
      <td>Burkina Faso</td>
      <td>Bana</td>
      <td>2014</td>
      <td>7</td>
      <td>11.233</td>
      <td>-4.472</td>
      <td>F</td>
      <td>AG1000G-BF-B</td>
      <td>v3</td>
      <td>0.986</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>coluzzii</td>
      <td>coluzzii</td>
    </tr>
  </tbody>
</table>
<p>283 rows × 17 columns</p>
</div></div></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">sample_id</span></code> column gives the sample identifier used throughout all Ag1000G analyses.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">country</span></code>, <code class="docutils literal notranslate"><span class="pre">location</span></code>, <code class="docutils literal notranslate"><span class="pre">latitude</span></code> and <code class="docutils literal notranslate"><span class="pre">longitude</span></code> columns give the location where the specimen was collected.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">year</span></code> and <code class="docutils literal notranslate"><span class="pre">month</span></code> columns give the approximate date when the specimen was collected.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">sex_call</span></code> column gives the gender as determined from the sequence data.</p>
<p>To load metadata for all wild-caught samples, you can use the shortcut “v3_wild”, e.g.:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_samples</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">(</span><span class="n">sample_sets</span><span class="o">=</span><span class="s2">&quot;v3_wild&quot;</span><span class="p">)</span>
<span class="n">df_samples</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample_id</th>
      <th>partner_sample_id</th>
      <th>contributor</th>
      <th>country</th>
      <th>location</th>
      <th>year</th>
      <th>month</th>
      <th>latitude</th>
      <th>longitude</th>
      <th>sex_call</th>
      <th>sample_set</th>
      <th>release</th>
      <th>aim_fraction_colu</th>
      <th>aim_fraction_arab</th>
      <th>species_gambcolu_arabiensis</th>
      <th>species_gambiae_coluzzii</th>
      <th>species</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AR0047-C</td>
      <td>LUA047</td>
      <td>Joao Pinto</td>
      <td>Angola</td>
      <td>Luanda</td>
      <td>2009</td>
      <td>4</td>
      <td>-8.884</td>
      <td>13.302</td>
      <td>F</td>
      <td>AG1000G-AO</td>
      <td>v3</td>
      <td>0.945</td>
      <td>0.001</td>
      <td>gamb_colu</td>
      <td>coluzzii</td>
      <td>coluzzii</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AR0049-C</td>
      <td>LUA049</td>
      <td>Joao Pinto</td>
      <td>Angola</td>
      <td>Luanda</td>
      <td>2009</td>
      <td>4</td>
      <td>-8.884</td>
      <td>13.302</td>
      <td>F</td>
      <td>AG1000G-AO</td>
      <td>v3</td>
      <td>0.933</td>
      <td>0.001</td>
      <td>gamb_colu</td>
      <td>coluzzii</td>
      <td>coluzzii</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AR0051-C</td>
      <td>LUA051</td>
      <td>Joao Pinto</td>
      <td>Angola</td>
      <td>Luanda</td>
      <td>2009</td>
      <td>4</td>
      <td>-8.884</td>
      <td>13.302</td>
      <td>F</td>
      <td>AG1000G-AO</td>
      <td>v3</td>
      <td>0.937</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>coluzzii</td>
      <td>coluzzii</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AR0061-C</td>
      <td>LUA061</td>
      <td>Joao Pinto</td>
      <td>Angola</td>
      <td>Luanda</td>
      <td>2009</td>
      <td>4</td>
      <td>-8.884</td>
      <td>13.302</td>
      <td>F</td>
      <td>AG1000G-AO</td>
      <td>v3</td>
      <td>0.938</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>coluzzii</td>
      <td>coluzzii</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AR0078-C</td>
      <td>LUA078</td>
      <td>Joao Pinto</td>
      <td>Angola</td>
      <td>Luanda</td>
      <td>2009</td>
      <td>4</td>
      <td>-8.884</td>
      <td>13.302</td>
      <td>F</td>
      <td>AG1000G-AO</td>
      <td>v3</td>
      <td>0.926</td>
      <td>0.001</td>
      <td>gamb_colu</td>
      <td>coluzzii</td>
      <td>coluzzii</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>2779</th>
      <td>AC0295-C</td>
      <td>K92</td>
      <td>Martin Donnelly</td>
      <td>Uganda</td>
      <td>Kihihi</td>
      <td>2012</td>
      <td>11</td>
      <td>-0.751</td>
      <td>29.701</td>
      <td>F</td>
      <td>AG1000G-UG</td>
      <td>v3</td>
      <td>0.026</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>2780</th>
      <td>AC0296-C</td>
      <td>K93</td>
      <td>Martin Donnelly</td>
      <td>Uganda</td>
      <td>Kihihi</td>
      <td>2012</td>
      <td>11</td>
      <td>-0.751</td>
      <td>29.701</td>
      <td>F</td>
      <td>AG1000G-UG</td>
      <td>v3</td>
      <td>0.029</td>
      <td>0.003</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>2781</th>
      <td>AC0297-C</td>
      <td>K94</td>
      <td>Martin Donnelly</td>
      <td>Uganda</td>
      <td>Kihihi</td>
      <td>2012</td>
      <td>11</td>
      <td>-0.751</td>
      <td>29.701</td>
      <td>F</td>
      <td>AG1000G-UG</td>
      <td>v3</td>
      <td>0.026</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>2782</th>
      <td>AC0298-C</td>
      <td>K95</td>
      <td>Martin Donnelly</td>
      <td>Uganda</td>
      <td>Kihihi</td>
      <td>2012</td>
      <td>11</td>
      <td>-0.751</td>
      <td>29.701</td>
      <td>F</td>
      <td>AG1000G-UG</td>
      <td>v3</td>
      <td>0.029</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
    <tr>
      <th>2783</th>
      <td>AC0299-C</td>
      <td>K96</td>
      <td>Martin Donnelly</td>
      <td>Uganda</td>
      <td>Kihihi</td>
      <td>2012</td>
      <td>11</td>
      <td>-0.751</td>
      <td>29.701</td>
      <td>F</td>
      <td>AG1000G-UG</td>
      <td>v3</td>
      <td>0.029</td>
      <td>0.002</td>
      <td>gamb_colu</td>
      <td>gambiae</td>
      <td>gambiae</td>
    </tr>
  </tbody>
</table>
<p>2784 rows × 17 columns</p>
</div></div></div>
</div>
<p>Pandas can be used to explore and query the sample metadata in various ways. E.g., here is a summary of the numbers of samples by species:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_samples</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;species&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>species
arabiensis                          368
coluzzii                            675
gambiae                            1571
intermediate_arabiensis_gambiae       1
intermediate_gambiae_coluzzii       169
dtype: int64
</pre></div>
</div>
</div>
</div>
<p>Note that samples within a sample set may belong to different species. For convenience, we have made a species call for all samples in Ag1000G phase 3, using the genomic data. Calling species is not always straightforward, and we have used two different methods for species calling, ancestry informative markers (AIM) and principal components analysis (PCA). When loading the sample metadata, the AIM species calls are included by default. The results of these two different methods generally agree, although there are some populations where results are different, particularly in Guinea-Bissau, The Gambia, Kenya and Tanzania. If you have any questions about how to interpret these species calls, please get in touch.</p>
</div>
<div class="section" id="snp-sites-and-alleles">
<h2>SNP sites and alleles<a class="headerlink" href="#snp-sites-and-alleles" title="Permalink to this headline">¶</a></h2>
<p>We have called SNP genotypes in all samples at all positions in the genome where the reference allele is not “N”. Data on this set of genomic positions and alleles for a given chromosome arm (e.g., 3R) can be accessed as dask arrays as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pos</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">snp_sites</span><span class="p">(</span><span class="n">contig</span><span class="o">=</span><span class="s2">&quot;3R&quot;</span><span class="p">)</span>
<span class="n">pos</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 208.91 MB </td> <td> 134.22 MB </td></tr>
    <tr><th> Shape </th><td> (52226568,) </td> <td> (33554432,) </td></tr>
    <tr><th> Count </th><td> 2 Tasks </td><td> 2 Chunks </td></tr>
    <tr><th> Type </th><td> int32 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="77" y1="0" x2="77" y2="25" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >52226568</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
</td>
</tr>
</table></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ref</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 52.23 MB </td> <td> 52.23 MB </td></tr>
    <tr><th> Shape </th><td> (52226568,) </td> <td> (52226568,) </td></tr>
    <tr><th> Count </th><td> 1 Tasks </td><td> 1 Chunks </td></tr>
    <tr><th> Type </th><td> |S1 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >52226568</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
</td>
</tr>
</table></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">alt</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 156.68 MB </td> <td> 133.69 MB </td></tr>
    <tr><th> Shape </th><td> (52226568, 3) </td> <td> (44564480, 3) </td></tr>
    <tr><th> Count </th><td> 2 Tasks </td><td> 2 Chunks </td></tr>
    <tr><th> Type </th><td> |S1 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="75" height="170" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="25" y2="0" style="stroke-width:2" />
  <line x1="0" y1="102" x2="25" y2="102" />
  <line x1="0" y1="120" x2="25" y2="120" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="120" style="stroke-width:2" />
  <line x1="25" y1="0" x2="25" y2="120" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 25.412616514582485,0.0 25.412616514582485,120.0 0.0,120.0" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="12.706308" y="140.000000" font-size="1.0rem" font-weight="100" text-anchor="middle" >3</text>
  <text x="45.412617" y="60.000000" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,45.412617,60.000000)">52226568</text>
</svg>
</td>
</tr>
</table></div></div>
</div>
<p>Data can be loaded into memory as numpy arrays as shown in the following examples.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># read first 10 SNP positions into a numpy array</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">pos</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10], dtype=int32)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># read first 10 SNP reference alleles</span>
<span class="n">r</span> <span class="o">=</span> <span class="n">ref</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
<span class="n">r</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([b&#39;C&#39;, b&#39;C&#39;, b&#39;T&#39;, b&#39;C&#39;, b&#39;T&#39;, b&#39;A&#39;, b&#39;C&#39;, b&#39;G&#39;, b&#39;T&#39;, b&#39;T&#39;],
      dtype=&#39;|S1&#39;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># read first 10 SNP alternate alleles</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">alt</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
<span class="n">a</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[b&#39;A&#39;, b&#39;T&#39;, b&#39;G&#39;],
       [b&#39;A&#39;, b&#39;T&#39;, b&#39;G&#39;],
       [b&#39;A&#39;, b&#39;C&#39;, b&#39;G&#39;],
       [b&#39;A&#39;, b&#39;T&#39;, b&#39;G&#39;],
       [b&#39;A&#39;, b&#39;C&#39;, b&#39;G&#39;],
       [b&#39;C&#39;, b&#39;T&#39;, b&#39;G&#39;],
       [b&#39;A&#39;, b&#39;T&#39;, b&#39;G&#39;],
       [b&#39;A&#39;, b&#39;C&#39;, b&#39;T&#39;],
       [b&#39;A&#39;, b&#39;C&#39;, b&#39;G&#39;],
       [b&#39;A&#39;, b&#39;C&#39;, b&#39;G&#39;]], dtype=&#39;|S1&#39;)
</pre></div>
</div>
</div>
</div>
<p>Note that we have chosen to genotype all samples at all sites in the genome, assuming all possible SNP alleles. Not all of these alternate alleles will actually have been observed in the Ag3 samples. To determine which sites and alleles are segregating, an allele count can be performed over the samples you are interested in. See the example below.</p>
</div>
<div class="section" id="site-filters">
<h2>Site filters<a class="headerlink" href="#site-filters" title="Permalink to this headline">¶</a></h2>
<p>SNP calling is not always reliable, and we have created some site filters to allow excluding low quality SNPs.</p>
<p>Because different species may have different genome accessibility issues, we have created three separate site filters:</p>
<ul class="simple">
<li><p>The “gamb_colu” filter is design for working only with <em>An. gambiae</em> and/or <em>An. coluzzii</em> samples.</p></li>
<li><p>The “arab” filter is designed for working with <em>An. arabiensis</em> samples.</p></li>
<li><p>The “gamb_colu_arab” filter is suitable for when analysing samples of any species together.</p></li>
</ul>
<p>Each set of site filters provides a “filter_pass” Boolean mask for each chromosome arm, where True indicates that the site passed the filter and is accessible to high quality SNP calling.</p>
<p>The site filters data can be accessed as dask arrays as shown in the examples below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># access gamb_colu_arab site filters for chromosome arm 3R as a dask array</span>
<span class="n">filter_pass</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">site_filters</span><span class="p">(</span><span class="n">contig</span><span class="o">=</span><span class="s2">&quot;3R&quot;</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="s2">&quot;gamb_colu_arab&quot;</span><span class="p">)</span>
<span class="n">filter_pass</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 52.23 MB </td> <td> 52.23 MB </td></tr>
    <tr><th> Shape </th><td> (52226568,) </td> <td> (52226568,) </td></tr>
    <tr><th> Count </th><td> 1 Tasks </td><td> 1 Chunks </td></tr>
    <tr><th> Type </th><td> bool </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >52226568</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
</td>
</tr>
</table></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># read filter values for first 10 SNPs (True means the site passes filters)</span>
<span class="n">f</span> <span class="o">=</span> <span class="n">filter_pass</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
<span class="n">f</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([False, False, False, False, False, False, False, False, False,
       False])
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="snp-genotypes">
<h2>SNP genotypes<a class="headerlink" href="#snp-genotypes" title="Permalink to this headline">¶</a></h2>
<p>SNP genotypes for individual samples are available. Genotypes are stored as a three-dimensional array, where the first dimension corresponds to genomic positions, the second dimension is samples, and the third dimension is ploidy (2). Values coded as integers, where -1 represents a missing value, 0 represents the reference allele, and 1, 2, and 3 represent alternate alleles.</p>
<p>SNP genotypes can be accessed as dask arrays as shown below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gt</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">snp_genotypes</span><span class="p">(</span><span class="n">contig</span><span class="o">=</span><span class="s2">&quot;3R&quot;</span><span class="p">,</span> <span class="n">sample_sets</span><span class="o">=</span><span class="s2">&quot;v3_wild&quot;</span><span class="p">)</span>
<span class="n">gt</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 290.80 GB </td> <td> 122.40 MB </td></tr>
    <tr><th> Shape </th><td> (52226568, 2784, 2) </td> <td> (600000, 102, 2) </td></tr>
    <tr><th> Count </th><td> 16774 Tasks </td><td> 5032 Chunks </td></tr>
    <tr><th> Type </th><td> int8 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="156" height="146" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="10" y1="0" x2="80" y2="71" />
  <line x1="10" y1="2" x2="80" y2="72" />
  <line x1="10" y1="3" x2="80" y2="74" />
  <line x1="10" y1="4" x2="80" y2="75" />
  <line x1="10" y1="6" x2="80" y2="77" />
  <line x1="10" y1="8" x2="80" y2="78" />
  <line x1="10" y1="9" x2="80" y2="79" />
  <line x1="10" y1="9" x2="80" y2="80" />
  <line x1="10" y1="11" x2="80" y2="81" />
  <line x1="10" y1="12" x2="80" y2="82" />
  <line x1="10" y1="13" x2="80" y2="84" />
  <line x1="10" y1="15" x2="80" y2="85" />
  <line x1="10" y1="16" x2="80" y2="86" />
  <line x1="10" y1="17" x2="80" y2="88" />
  <line x1="10" y1="18" x2="80" y2="89" />
  <line x1="10" y1="20" x2="80" y2="90" />
  <line x1="10" y1="21" x2="80" y2="92" />
  <line x1="10" y1="23" x2="80" y2="94" />
  <line x1="10" y1="25" x2="80" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="10" y2="25" style="stroke-width:2" />
  <line x1="13" y1="3" x2="13" y2="28" />
  <line x1="16" y1="6" x2="16" y2="32" />
  <line x1="20" y1="10" x2="20" y2="35" />
  <line x1="24" y1="14" x2="24" y2="39" />
  <line x1="28" y1="18" x2="28" y2="43" />
  <line x1="31" y1="21" x2="31" y2="47" />
  <line x1="35" y1="25" x2="35" y2="51" />
  <line x1="40" y1="30" x2="40" y2="55" />
  <line x1="43" y1="33" x2="43" y2="58" />
  <line x1="46" y1="36" x2="46" y2="62" />
  <line x1="50" y1="40" x2="50" y2="65" />
  <line x1="54" y1="44" x2="54" y2="70" />
  <line x1="58" y1="48" x2="58" y2="73" />
  <line x1="61" y1="51" x2="61" y2="77" />
  <line x1="65" y1="55" x2="65" y2="80" />
  <line x1="69" y1="59" x2="69" y2="84" />
  <line x1="72" y1="62" x2="72" y2="88" />
  <line x1="76" y1="66" x2="76" y2="91" />
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.0,0.0 80.58823529411765,70.58823529411765 80.58823529411765,96.00085180870013 10.0,25.41261651458248" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="35" y2="0" style="stroke-width:2" />
  <line x1="13" y1="3" x2="38" y2="3" />
  <line x1="16" y1="6" x2="42" y2="6" />
  <line x1="20" y1="10" x2="45" y2="10" />
  <line x1="24" y1="14" x2="49" y2="14" />
  <line x1="28" y1="18" x2="53" y2="18" />
  <line x1="31" y1="21" x2="57" y2="21" />
  <line x1="35" y1="25" x2="61" y2="25" />
  <line x1="40" y1="30" x2="65" y2="30" />
  <line x1="43" y1="33" x2="68" y2="33" />
  <line x1="46" y1="36" x2="72" y2="36" />
  <line x1="50" y1="40" x2="75" y2="40" />
  <line x1="54" y1="44" x2="80" y2="44" />
  <line x1="58" y1="48" x2="83" y2="48" />
  <line x1="61" y1="51" x2="87" y2="51" />
  <line x1="65" y1="55" x2="90" y2="55" />
  <line x1="69" y1="59" x2="94" y2="59" />
  <line x1="72" y1="62" x2="98" y2="62" />
  <line x1="76" y1="66" x2="101" y2="66" />
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="35" y1="0" x2="106" y2="70" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.0,0.0 35.41261651458248,0.0 106.00085180870013,70.58823529411765 80.58823529411765,70.58823529411765" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />
  <line x1="80" y1="71" x2="106" y2="71" />
  <line x1="80" y1="72" x2="106" y2="72" />
  <line x1="80" y1="74" x2="106" y2="74" />
  <line x1="80" y1="75" x2="106" y2="75" />
  <line x1="80" y1="77" x2="106" y2="77" />
  <line x1="80" y1="78" x2="106" y2="78" />
  <line x1="80" y1="79" x2="106" y2="79" />
  <line x1="80" y1="80" x2="106" y2="80" />
  <line x1="80" y1="81" x2="106" y2="81" />
  <line x1="80" y1="82" x2="106" y2="82" />
  <line x1="80" y1="84" x2="106" y2="84" />
  <line x1="80" y1="85" x2="106" y2="85" />
  <line x1="80" y1="86" x2="106" y2="86" />
  <line x1="80" y1="88" x2="106" y2="88" />
  <line x1="80" y1="89" x2="106" y2="89" />
  <line x1="80" y1="90" x2="106" y2="90" />
  <line x1="80" y1="92" x2="106" y2="92" />
  <line x1="80" y1="94" x2="106" y2="94" />
  <line x1="80" y1="96" x2="106" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />
  <line x1="106" y1="70" x2="106" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="80.58823529411765,70.58823529411765 106.00085180870013,70.58823529411765 106.00085180870013,96.00085180870013 80.58823529411765,96.00085180870013" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Text -->
  <text x="93.294544" y="116.000852" font-size="1.0rem" font-weight="100" text-anchor="middle" >2</text>
  <text x="126.000852" y="83.294544" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,126.000852,83.294544)">2784</text>
  <text x="35.294118" y="80.706734" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(45,35.294118,80.706734)">52226568</text>
</svg>
</td>
</tr>
</table></div></div>
</div>
<p>Note that the columns of this array (second dimension) match the rows in the sample metadata, if the same sample sets were loaded. I.e.:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_samples</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">(</span><span class="n">sample_sets</span><span class="o">=</span><span class="s2">&quot;v3_wild&quot;</span><span class="p">)</span>
<span class="n">gt</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">snp_genotypes</span><span class="p">(</span><span class="n">contig</span><span class="o">=</span><span class="s2">&quot;3R&quot;</span><span class="p">,</span> <span class="n">sample_sets</span><span class="o">=</span><span class="s2">&quot;v3_wild&quot;</span><span class="p">)</span>
<span class="nb">len</span><span class="p">(</span><span class="n">df_samples</span><span class="p">)</span> <span class="o">==</span> <span class="n">gt</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>True
</pre></div>
</div>
</div>
</div>
<p>You can use this correspondance to apply further subsetting operations to the genotypes by querying the sample metadata. E.g.:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">loc_gambiae</span> <span class="o">=</span> <span class="n">df_samples</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="s2">&quot;species == &#39;gambiae&#39;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;found </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">loc_gambiae</span><span class="p">)</span><span class="si">}</span><span class="s2"> gambiae samples&quot;</span><span class="p">)</span>
<span class="n">gt_gambiae</span> <span class="o">=</span> <span class="n">da</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">loc_gambiae</span><span class="p">,</span> <span class="n">gt</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">gt_gambiae</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>found 1571 gambiae samples
</pre></div>
</div>
<div class="output text_html"><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 164.10 GB </td> <td> 120.00 MB </td></tr>
    <tr><th> Shape </th><td> (52226568, 1571, 2) </td> <td> (600000, 100, 2) </td></tr>
    <tr><th> Count </th><td> 21262 Tasks </td><td> 4488 Chunks </td></tr>
    <tr><th> Type </th><td> int8 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="156" height="146" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="10" y1="0" x2="80" y2="71" />
  <line x1="10" y1="2" x2="80" y2="72" />
  <line x1="10" y1="3" x2="80" y2="74" />
  <line x1="10" y1="4" x2="80" y2="75" />
  <line x1="10" y1="7" x2="80" y2="78" />
  <line x1="10" y1="9" x2="80" y2="80" />
  <line x1="10" y1="11" x2="80" y2="81" />
  <line x1="10" y1="11" x2="80" y2="82" />
  <line x1="10" y1="13" x2="80" y2="84" />
  <line x1="10" y1="14" x2="80" y2="85" />
  <line x1="10" y1="15" x2="80" y2="85" />
  <line x1="10" y1="16" x2="80" y2="87" />
  <line x1="10" y1="17" x2="80" y2="88" />
  <line x1="10" y1="18" x2="80" y2="89" />
  <line x1="10" y1="19" x2="80" y2="90" />
  <line x1="10" y1="20" x2="80" y2="91" />
  <line x1="10" y1="21" x2="80" y2="92" />
  <line x1="10" y1="22" x2="80" y2="92" />
  <line x1="10" y1="25" x2="80" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="10" y2="25" style="stroke-width:2" />
  <line x1="13" y1="3" x2="13" y2="28" />
  <line x1="16" y1="6" x2="16" y2="32" />
  <line x1="20" y1="10" x2="20" y2="35" />
  <line x1="24" y1="14" x2="24" y2="39" />
  <line x1="28" y1="18" x2="28" y2="43" />
  <line x1="31" y1="21" x2="31" y2="47" />
  <line x1="35" y1="25" x2="35" y2="51" />
  <line x1="40" y1="30" x2="40" y2="55" />
  <line x1="43" y1="33" x2="43" y2="58" />
  <line x1="46" y1="36" x2="46" y2="62" />
  <line x1="50" y1="40" x2="50" y2="65" />
  <line x1="54" y1="44" x2="54" y2="70" />
  <line x1="58" y1="48" x2="58" y2="73" />
  <line x1="61" y1="51" x2="61" y2="77" />
  <line x1="65" y1="55" x2="65" y2="80" />
  <line x1="69" y1="59" x2="69" y2="84" />
  <line x1="72" y1="62" x2="72" y2="88" />
  <line x1="76" y1="66" x2="76" y2="91" />
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.0,0.0 80.58823529411765,70.58823529411765 80.58823529411765,96.00085180870013 10.0,25.41261651458249" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="35" y2="0" style="stroke-width:2" />
  <line x1="13" y1="3" x2="38" y2="3" />
  <line x1="16" y1="6" x2="42" y2="6" />
  <line x1="20" y1="10" x2="45" y2="10" />
  <line x1="24" y1="14" x2="49" y2="14" />
  <line x1="28" y1="18" x2="53" y2="18" />
  <line x1="31" y1="21" x2="57" y2="21" />
  <line x1="35" y1="25" x2="61" y2="25" />
  <line x1="40" y1="30" x2="65" y2="30" />
  <line x1="43" y1="33" x2="68" y2="33" />
  <line x1="46" y1="36" x2="72" y2="36" />
  <line x1="50" y1="40" x2="75" y2="40" />
  <line x1="54" y1="44" x2="80" y2="44" />
  <line x1="58" y1="48" x2="83" y2="48" />
  <line x1="61" y1="51" x2="87" y2="51" />
  <line x1="65" y1="55" x2="90" y2="55" />
  <line x1="69" y1="59" x2="94" y2="59" />
  <line x1="72" y1="62" x2="98" y2="62" />
  <line x1="76" y1="66" x2="101" y2="66" />
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="35" y1="0" x2="106" y2="70" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.0,0.0 35.41261651458248,0.0 106.00085180870013,70.58823529411765 80.58823529411765,70.58823529411765" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />
  <line x1="80" y1="71" x2="106" y2="71" />
  <line x1="80" y1="72" x2="106" y2="72" />
  <line x1="80" y1="74" x2="106" y2="74" />
  <line x1="80" y1="75" x2="106" y2="75" />
  <line x1="80" y1="78" x2="106" y2="78" />
  <line x1="80" y1="80" x2="106" y2="80" />
  <line x1="80" y1="81" x2="106" y2="81" />
  <line x1="80" y1="82" x2="106" y2="82" />
  <line x1="80" y1="84" x2="106" y2="84" />
  <line x1="80" y1="85" x2="106" y2="85" />
  <line x1="80" y1="85" x2="106" y2="85" />
  <line x1="80" y1="87" x2="106" y2="87" />
  <line x1="80" y1="88" x2="106" y2="88" />
  <line x1="80" y1="89" x2="106" y2="89" />
  <line x1="80" y1="90" x2="106" y2="90" />
  <line x1="80" y1="91" x2="106" y2="91" />
  <line x1="80" y1="92" x2="106" y2="92" />
  <line x1="80" y1="92" x2="106" y2="92" />
  <line x1="80" y1="96" x2="106" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />
  <line x1="106" y1="70" x2="106" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="80.58823529411765,70.58823529411765 106.00085180870013,70.58823529411765 106.00085180870013,96.00085180870013 80.58823529411765,96.00085180870013" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Text -->
  <text x="93.294544" y="116.000852" font-size="1.0rem" font-weight="100" text-anchor="middle" >2</text>
  <text x="126.000852" y="83.294544" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,126.000852,83.294544)">1571</text>
  <text x="35.294118" y="80.706734" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(45,35.294118,80.706734)">52226568</text>
</svg>
</td>
</tr>
</table></div></div>
</div>
<p>Data can be read into memory as numpy arrays, e.g., read genotypes for the first 5 SNPs and the first 3 samples:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">g</span> <span class="o">=</span> <span class="n">gt</span><span class="p">[:</span><span class="mi">5</span><span class="p">,</span> <span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
<span class="n">g</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[[0, 0],
        [0, 0],
        [0, 0]],

       [[0, 0],
        [0, 0],
        [0, 0]],

       [[0, 0],
        [0, 0],
        [0, 0]],

       [[0, 0],
        [0, 0],
        [0, 0]],

       [[0, 0],
        [0, 0],
        [0, 0]]], dtype=int8)
</pre></div>
</div>
</div>
</div>
<p>If you want to work with the genotype calls, you may find it convenient to use <a class="reference external" href="http://scikit-allel.readthedocs.org/">scikit-allel</a>.
E.g., the code below sets up a genotype array.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># use the scikit-allel wrapper class for genotype calls</span>
<span class="n">gt</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">snp_genotypes</span><span class="p">(</span><span class="n">contig</span><span class="o">=</span><span class="s2">&quot;3R&quot;</span><span class="p">,</span> <span class="n">sample_sets</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;AG1000G-BF-A&quot;</span><span class="p">,</span> <span class="s2">&quot;AG1000G-BF-B&quot;</span><span class="p">])</span>
<span class="n">gt</span> <span class="o">=</span> <span class="n">allel</span><span class="o">.</span><span class="n">GenotypeDaskArray</span><span class="p">(</span><span class="n">gt</span><span class="p">)</span>
<span class="n">gt</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="allel allel-DisplayAs2D"><span>&lt;GenotypeDaskArray shape=(52226568, 283, 2) dtype=int8&gt;</span><table><thead><tr><th></th><th style="text-align: center">0</th><th style="text-align: center">1</th><th style="text-align: center">2</th><th style="text-align: center">3</th><th style="text-align: center">4</th><th style="text-align: center">...</th><th style="text-align: center">278</th><th style="text-align: center">279</th><th style="text-align: center">280</th><th style="text-align: center">281</th><th style="text-align: center">282</th></tr></thead><tbody><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">0</th><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">...</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">1</th><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">...</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">2</th><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">...</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">...</th><td style="text-align: center" colspan="12">...</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">52226565</th><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">...</td><td style="text-align: center">0/0</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">52226566</th><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">...</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">52226567</th><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">...</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td></tr></tbody></table></div></div></div>
</div>
</div>
<div class="section" id="snp-calls-dataset">
<h2>SNP calls dataset<a class="headerlink" href="#snp-calls-dataset" title="Permalink to this headline">¶</a></h2>
<p>If you are familiar with <a class="reference external" href="http://xarray.pydata.org/en/stable/">xarray</a>, you may find it more convenient to access an xarray dataset which contains all of the data relating to SNP calls, including the SNP positions, alleles, site filters, and genotypes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ds_snps</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">snp_calls</span><span class="p">(</span><span class="n">contig</span><span class="o">=</span><span class="s2">&quot;3R&quot;</span><span class="p">,</span> <span class="n">sample_sets</span><span class="o">=</span><span class="s2">&quot;v3_wild&quot;</span><span class="p">)</span>
<span class="n">ds_snps</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div><svg style="position: absolute; width: 0; height: 0; overflow: hidden">
<defs>
<symbol id="icon-database" viewBox="0 0 32 32">
<path d="M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z"></path>
<path d="M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
<path d="M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
</symbol>
<symbol id="icon-file-text2" viewBox="0 0 32 32">
<path d="M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z"></path>
<path d="M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
</symbol>
</defs>
</svg>
<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.
 *
 */

:root {
  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));
  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));
  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));
  --xr-border-color: var(--jp-border-color2, #e0e0e0);
  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);
  --xr-background-color: var(--jp-layout-color0, white);
  --xr-background-color-row-even: var(--jp-layout-color1, white);
  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);
}

html[theme=dark],
body.vscode-dark {
  --xr-font-color0: rgba(255, 255, 255, 1);
  --xr-font-color2: rgba(255, 255, 255, 0.54);
  --xr-font-color3: rgba(255, 255, 255, 0.38);
  --xr-border-color: #1F1F1F;
  --xr-disabled-color: #515151;
  --xr-background-color: #111111;
  --xr-background-color-row-even: #111111;
  --xr-background-color-row-odd: #313131;
}

.xr-wrap {
  display: block;
  min-width: 300px;
  max-width: 700px;
}

.xr-text-repr-fallback {
  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */
  display: none;
}

.xr-header {
  padding-top: 6px;
  padding-bottom: 6px;
  margin-bottom: 4px;
  border-bottom: solid 1px var(--xr-border-color);
}

.xr-header > div,
.xr-header > ul {
  display: inline;
  margin-top: 0;
  margin-bottom: 0;
}

.xr-obj-type,
.xr-array-name {
  margin-left: 2px;
  margin-right: 10px;
}

.xr-obj-type {
  color: var(--xr-font-color2);
}

.xr-sections {
  padding-left: 0 !important;
  display: grid;
  grid-template-columns: 150px auto auto 1fr 20px 20px;
}

.xr-section-item {
  display: contents;
}

.xr-section-item input {
  display: none;
}

.xr-section-item input + label {
  color: var(--xr-disabled-color);
}

.xr-section-item input:enabled + label {
  cursor: pointer;
  color: var(--xr-font-color2);
}

.xr-section-item input:enabled + label:hover {
  color: var(--xr-font-color0);
}

.xr-section-summary {
  grid-column: 1;
  color: var(--xr-font-color2);
  font-weight: 500;
}

.xr-section-summary > span {
  display: inline-block;
  padding-left: 0.5em;
}

.xr-section-summary-in:disabled + label {
  color: var(--xr-font-color2);
}

.xr-section-summary-in + label:before {
  display: inline-block;
  content: '►';
  font-size: 11px;
  width: 15px;
  text-align: center;
}

.xr-section-summary-in:disabled + label:before {
  color: var(--xr-disabled-color);
}

.xr-section-summary-in:checked + label:before {
  content: '▼';
}

.xr-section-summary-in:checked + label > span {
  display: none;
}

.xr-section-summary,
.xr-section-inline-details {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xr-section-inline-details {
  grid-column: 2 / -1;
}

.xr-section-details {
  display: none;
  grid-column: 1 / -1;
  margin-bottom: 5px;
}

.xr-section-summary-in:checked ~ .xr-section-details {
  display: contents;
}

.xr-array-wrap {
  grid-column: 1 / -1;
  display: grid;
  grid-template-columns: 20px auto;
}

.xr-array-wrap > label {
  grid-column: 1;
  vertical-align: top;
}

.xr-preview {
  color: var(--xr-font-color3);
}

.xr-array-preview,
.xr-array-data {
  padding: 0 5px !important;
  grid-column: 2;
}

.xr-array-data,
.xr-array-in:checked ~ .xr-array-preview {
  display: none;
}

.xr-array-in:checked ~ .xr-array-data,
.xr-array-preview {
  display: inline-block;
}

.xr-dim-list {
  display: inline-block !important;
  list-style: none;
  padding: 0 !important;
  margin: 0;
}

.xr-dim-list li {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.xr-dim-list:before {
  content: '(';
}

.xr-dim-list:after {
  content: ')';
}

.xr-dim-list li:not(:last-child):after {
  content: ',';
  padding-right: 5px;
}

.xr-has-index {
  font-weight: bold;
}

.xr-var-list,
.xr-var-item {
  display: contents;
}

.xr-var-item > div,
.xr-var-item label,
.xr-var-item > .xr-var-name span {
  background-color: var(--xr-background-color-row-even);
  margin-bottom: 0;
}

.xr-var-item > .xr-var-name:hover span {
  padding-right: 5px;
}

.xr-var-list > li:nth-child(odd) > div,
.xr-var-list > li:nth-child(odd) > label,
.xr-var-list > li:nth-child(odd) > .xr-var-name span {
  background-color: var(--xr-background-color-row-odd);
}

.xr-var-name {
  grid-column: 1;
}

.xr-var-dims {
  grid-column: 2;
}

.xr-var-dtype {
  grid-column: 3;
  text-align: right;
  color: var(--xr-font-color2);
}

.xr-var-preview {
  grid-column: 4;
}

.xr-var-name,
.xr-var-dims,
.xr-var-dtype,
.xr-preview,
.xr-attrs dt {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-right: 10px;
}

.xr-var-name:hover,
.xr-var-dims:hover,
.xr-var-dtype:hover,
.xr-attrs dt:hover {
  overflow: visible;
  width: auto;
  z-index: 1;
}

.xr-var-attrs,
.xr-var-data {
  display: none;
  background-color: var(--xr-background-color) !important;
  padding-bottom: 5px !important;
}

.xr-var-attrs-in:checked ~ .xr-var-attrs,
.xr-var-data-in:checked ~ .xr-var-data {
  display: block;
}

.xr-var-data > table {
  float: right;
}

.xr-var-name span,
.xr-var-data,
.xr-attrs {
  padding-left: 25px !important;
}

.xr-attrs,
.xr-var-attrs,
.xr-var-data {
  grid-column: 1 / -1;
}

dl.xr-attrs {
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: 125px auto;
}

.xr-attrs dt,
.xr-attrs dd {
  padding: 0;
  margin: 0;
  float: left;
  padding-right: 10px;
  width: auto;
}

.xr-attrs dt {
  font-weight: normal;
  grid-column: 1;
}

.xr-attrs dt:hover span {
  display: inline-block;
  background: var(--xr-background-color);
  padding-right: 10px;
}

.xr-attrs dd {
  grid-column: 2;
  white-space: pre-wrap;
  word-break: break-all;
}

.xr-icon-database,
.xr-icon-file-text2 {
  display: inline-block;
  vertical-align: middle;
  width: 1em;
  height: 1.5em !important;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
}
</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;
Dimensions:                             (alleles: 4, ploidy: 2, samples: 2784, variants: 52226568)
Coordinates:
    variant_position                    (variants) int32 dask.array&lt;chunksize=(33554432,), meta=np.ndarray&gt;
    variant_contig                      (variants) uint8 dask.array&lt;chunksize=(33554432,), meta=np.ndarray&gt;
    sample_id                           (samples) |S24 dask.array&lt;chunksize=(81,), meta=np.ndarray&gt;
Dimensions without coordinates: alleles, ploidy, samples, variants
Data variables:
    variant_allele                      (variants, alleles) |S1 dask.array&lt;chunksize=(44564480, 1), meta=np.ndarray&gt;
    variant_filter_pass_gamb_colu_arab  (variants) bool dask.array&lt;chunksize=(52226568,), meta=np.ndarray&gt;
    variant_filter_pass_gamb_colu       (variants) bool dask.array&lt;chunksize=(52226568,), meta=np.ndarray&gt;
    variant_filter_pass_arab            (variants) bool dask.array&lt;chunksize=(52226568,), meta=np.ndarray&gt;
    call_genotype                       (variants, samples, ploidy) int8 dask.array&lt;chunksize=(600000, 81, 2), meta=np.ndarray&gt;
    call_GQ                             (variants, samples) int16 dask.array&lt;chunksize=(600000, 81), meta=np.ndarray&gt;
    call_MQ                             (variants, samples) int16 dask.array&lt;chunksize=(600000, 81), meta=np.ndarray&gt;
    call_AD                             (variants, samples, alleles) int16 dask.array&lt;chunksize=(300000, 50, 4), meta=np.ndarray&gt;
    call_genotype_mask                  (variants, samples, ploidy) bool dask.array&lt;chunksize=(600000, 81, 2), meta=np.ndarray&gt;
Attributes:
    contigs:  (&#x27;2R&#x27;, &#x27;2L&#x27;, &#x27;3R&#x27;, &#x27;3L&#x27;, &#x27;X&#x27;)</pre><div class='xr-wrap' hidden><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-7cdb1060-2444-49ab-99cf-6ac5005c8d5d' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-7cdb1060-2444-49ab-99cf-6ac5005c8d5d' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>alleles</span>: 4</li><li><span>ploidy</span>: 2</li><li><span>samples</span>: 2784</li><li><span>variants</span>: 52226568</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-8f5f6a90-de45-45c1-9f97-83bd08f39b9c' class='xr-section-summary-in' type='checkbox'  checked><label for='section-8f5f6a90-de45-45c1-9f97-83bd08f39b9c' class='xr-section-summary' >Coordinates: <span>(3)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>variant_position</span></div><div class='xr-var-dims'>(variants)</div><div class='xr-var-dtype'>int32</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(33554432,), meta=np.ndarray&gt;</div><input id='attrs-8b569c10-c5ef-462a-8fd8-25a1c0550520' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8b569c10-c5ef-462a-8fd8-25a1c0550520' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-6242118b-3654-4441-a146-e7b272d20ee4' class='xr-var-data-in' type='checkbox'><label for='data-6242118b-3654-4441-a146-e7b272d20ee4' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 208.91 MB </td> <td> 134.22 MB </td></tr>
    <tr><th> Shape </th><td> (52226568,) </td> <td> (33554432,) </td></tr>
    <tr><th> Count </th><td> 2 Tasks </td><td> 2 Chunks </td></tr>
    <tr><th> Type </th><td> int32 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="77" y1="0" x2="77" y2="25" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >52226568</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
</td>
</tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>variant_contig</span></div><div class='xr-var-dims'>(variants)</div><div class='xr-var-dtype'>uint8</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(33554432,), meta=np.ndarray&gt;</div><input id='attrs-0056f6d5-dd08-4453-b035-0387215af6ad' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0056f6d5-dd08-4453-b035-0387215af6ad' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5dfe9c1a-b9c1-403b-ba4d-f4f9a8b42479' class='xr-var-data-in' type='checkbox'><label for='data-5dfe9c1a-b9c1-403b-ba4d-f4f9a8b42479' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 52.23 MB </td> <td> 33.55 MB </td></tr>
    <tr><th> Shape </th><td> (52226568,) </td> <td> (33554432,) </td></tr>
    <tr><th> Count </th><td> 2 Tasks </td><td> 2 Chunks </td></tr>
    <tr><th> Type </th><td> uint8 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="77" y1="0" x2="77" y2="25" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >52226568</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
</td>
</tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>sample_id</span></div><div class='xr-var-dims'>(samples)</div><div class='xr-var-dtype'>|S24</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(81,), meta=np.ndarray&gt;</div><input id='attrs-ea65d085-75ca-46d4-8f19-1fcd342016eb' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ea65d085-75ca-46d4-8f19-1fcd342016eb' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-99b93441-9730-425e-b5bd-eadf21e2c057' class='xr-var-data-in' type='checkbox'><label for='data-99b93441-9730-425e-b5bd-eadf21e2c057' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 66.82 kB </td> <td> 7.27 kB </td></tr>
    <tr><th> Shape </th><td> (2784,) </td> <td> (303,) </td></tr>
    <tr><th> Count </th><td> 54 Tasks </td><td> 27 Chunks </td></tr>
    <tr><th> Type </th><td> |S24 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="3" y1="0" x2="3" y2="25" />
  <line x1="11" y1="0" x2="11" y2="25" />
  <line x1="16" y1="0" x2="16" y2="25" />
  <line x1="19" y1="0" x2="19" y2="25" />
  <line x1="26" y1="0" x2="26" y2="25" />
  <line x1="39" y1="0" x2="39" y2="25" />
  <line x1="43" y1="0" x2="43" y2="25" />
  <line x1="46" y1="0" x2="46" y2="25" />
  <line x1="49" y1="0" x2="49" y2="25" />
  <line x1="56" y1="0" x2="56" y2="25" />
  <line x1="58" y1="0" x2="58" y2="25" />
  <line x1="67" y1="0" x2="67" y2="25" />
  <line x1="75" y1="0" x2="75" y2="25" />
  <line x1="75" y1="0" x2="75" y2="25" />
  <line x1="83" y1="0" x2="83" y2="25" />
  <line x1="86" y1="0" x2="86" y2="25" />
  <line x1="91" y1="0" x2="91" y2="25" />
  <line x1="94" y1="0" x2="94" y2="25" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >2784</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
</td>
</tr>
</table></div></li></ul></div></li><li class='xr-section-item'><input id='section-e65f92df-161e-461f-bf64-1d5343c43c37' class='xr-section-summary-in' type='checkbox'  checked><label for='section-e65f92df-161e-461f-bf64-1d5343c43c37' class='xr-section-summary' >Data variables: <span>(9)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>variant_allele</span></div><div class='xr-var-dims'>(variants, alleles)</div><div class='xr-var-dtype'>|S1</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(44564480, 1), meta=np.ndarray&gt;</div><input id='attrs-9538ff49-a8ce-439b-9404-533629088679' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-9538ff49-a8ce-439b-9404-533629088679' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a916505d-c12b-4583-aeaf-8ffa2fbf79e1' class='xr-var-data-in' type='checkbox'><label for='data-a916505d-c12b-4583-aeaf-8ffa2fbf79e1' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 208.91 MB </td> <td> 133.69 MB </td></tr>
    <tr><th> Shape </th><td> (52226568, 4) </td> <td> (44564480, 3) </td></tr>
    <tr><th> Count </th><td> 12 Tasks </td><td> 4 Chunks </td></tr>
    <tr><th> Type </th><td> |S1 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="75" height="170" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="25" y2="0" style="stroke-width:2" />
  <line x1="0" y1="102" x2="25" y2="102" />
  <line x1="0" y1="120" x2="25" y2="120" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="120" style="stroke-width:2" />
  <line x1="6" y1="0" x2="6" y2="120" />
  <line x1="25" y1="0" x2="25" y2="120" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 25.412616514582485,0.0 25.412616514582485,120.0 0.0,120.0" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="12.706308" y="140.000000" font-size="1.0rem" font-weight="100" text-anchor="middle" >4</text>
  <text x="45.412617" y="60.000000" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,45.412617,60.000000)">52226568</text>
</svg>
</td>
</tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>variant_filter_pass_gamb_colu_arab</span></div><div class='xr-var-dims'>(variants)</div><div class='xr-var-dtype'>bool</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(52226568,), meta=np.ndarray&gt;</div><input id='attrs-04afcc4b-8a29-4974-9d15-022bcbf3eac2' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-04afcc4b-8a29-4974-9d15-022bcbf3eac2' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-f3f505be-95ba-4745-8697-f879ccdee3cd' class='xr-var-data-in' type='checkbox'><label for='data-f3f505be-95ba-4745-8697-f879ccdee3cd' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 52.23 MB </td> <td> 52.23 MB </td></tr>
    <tr><th> Shape </th><td> (52226568,) </td> <td> (52226568,) </td></tr>
    <tr><th> Count </th><td> 1 Tasks </td><td> 1 Chunks </td></tr>
    <tr><th> Type </th><td> bool </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >52226568</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
</td>
</tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>variant_filter_pass_gamb_colu</span></div><div class='xr-var-dims'>(variants)</div><div class='xr-var-dtype'>bool</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(52226568,), meta=np.ndarray&gt;</div><input id='attrs-83b80be9-206a-468c-a3f9-92b0b384605d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-83b80be9-206a-468c-a3f9-92b0b384605d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5d011e4a-aa9c-4664-9e13-62e2203e4de2' class='xr-var-data-in' type='checkbox'><label for='data-5d011e4a-aa9c-4664-9e13-62e2203e4de2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 52.23 MB </td> <td> 52.23 MB </td></tr>
    <tr><th> Shape </th><td> (52226568,) </td> <td> (52226568,) </td></tr>
    <tr><th> Count </th><td> 1 Tasks </td><td> 1 Chunks </td></tr>
    <tr><th> Type </th><td> bool </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >52226568</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
</td>
</tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>variant_filter_pass_arab</span></div><div class='xr-var-dims'>(variants)</div><div class='xr-var-dtype'>bool</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(52226568,), meta=np.ndarray&gt;</div><input id='attrs-5751ec0d-0e2b-4207-927c-91691e246ae3' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5751ec0d-0e2b-4207-927c-91691e246ae3' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1bd7647d-f460-4268-861e-b3f845574654' class='xr-var-data-in' type='checkbox'><label for='data-1bd7647d-f460-4268-861e-b3f845574654' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 52.23 MB </td> <td> 52.23 MB </td></tr>
    <tr><th> Shape </th><td> (52226568,) </td> <td> (52226568,) </td></tr>
    <tr><th> Count </th><td> 1 Tasks </td><td> 1 Chunks </td></tr>
    <tr><th> Type </th><td> bool </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >52226568</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
</td>
</tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>call_genotype</span></div><div class='xr-var-dims'>(variants, samples, ploidy)</div><div class='xr-var-dtype'>int8</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(600000, 81, 2), meta=np.ndarray&gt;</div><input id='attrs-13693214-5e86-4421-a7dd-67f73620016d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-13693214-5e86-4421-a7dd-67f73620016d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9178a3b9-1bce-4534-b962-57b473470bf0' class='xr-var-data-in' type='checkbox'><label for='data-9178a3b9-1bce-4534-b962-57b473470bf0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 290.80 GB </td> <td> 122.40 MB </td></tr>
    <tr><th> Shape </th><td> (52226568, 2784, 2) </td> <td> (600000, 102, 2) </td></tr>
    <tr><th> Count </th><td> 16774 Tasks </td><td> 5032 Chunks </td></tr>
    <tr><th> Type </th><td> int8 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="156" height="146" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="10" y1="0" x2="80" y2="71" />
  <line x1="10" y1="2" x2="80" y2="72" />
  <line x1="10" y1="3" x2="80" y2="74" />
  <line x1="10" y1="4" x2="80" y2="75" />
  <line x1="10" y1="6" x2="80" y2="77" />
  <line x1="10" y1="8" x2="80" y2="78" />
  <line x1="10" y1="9" x2="80" y2="79" />
  <line x1="10" y1="9" x2="80" y2="80" />
  <line x1="10" y1="11" x2="80" y2="81" />
  <line x1="10" y1="12" x2="80" y2="82" />
  <line x1="10" y1="13" x2="80" y2="84" />
  <line x1="10" y1="15" x2="80" y2="85" />
  <line x1="10" y1="16" x2="80" y2="86" />
  <line x1="10" y1="17" x2="80" y2="88" />
  <line x1="10" y1="18" x2="80" y2="89" />
  <line x1="10" y1="20" x2="80" y2="90" />
  <line x1="10" y1="21" x2="80" y2="92" />
  <line x1="10" y1="23" x2="80" y2="94" />
  <line x1="10" y1="25" x2="80" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="10" y2="25" style="stroke-width:2" />
  <line x1="13" y1="3" x2="13" y2="28" />
  <line x1="16" y1="6" x2="16" y2="32" />
  <line x1="20" y1="10" x2="20" y2="35" />
  <line x1="24" y1="14" x2="24" y2="39" />
  <line x1="28" y1="18" x2="28" y2="43" />
  <line x1="31" y1="21" x2="31" y2="47" />
  <line x1="35" y1="25" x2="35" y2="51" />
  <line x1="40" y1="30" x2="40" y2="55" />
  <line x1="43" y1="33" x2="43" y2="58" />
  <line x1="46" y1="36" x2="46" y2="62" />
  <line x1="50" y1="40" x2="50" y2="65" />
  <line x1="54" y1="44" x2="54" y2="70" />
  <line x1="58" y1="48" x2="58" y2="73" />
  <line x1="61" y1="51" x2="61" y2="77" />
  <line x1="65" y1="55" x2="65" y2="80" />
  <line x1="69" y1="59" x2="69" y2="84" />
  <line x1="72" y1="62" x2="72" y2="88" />
  <line x1="76" y1="66" x2="76" y2="91" />
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.0,0.0 80.58823529411765,70.58823529411765 80.58823529411765,96.00085180870013 10.0,25.41261651458248" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="35" y2="0" style="stroke-width:2" />
  <line x1="13" y1="3" x2="38" y2="3" />
  <line x1="16" y1="6" x2="42" y2="6" />
  <line x1="20" y1="10" x2="45" y2="10" />
  <line x1="24" y1="14" x2="49" y2="14" />
  <line x1="28" y1="18" x2="53" y2="18" />
  <line x1="31" y1="21" x2="57" y2="21" />
  <line x1="35" y1="25" x2="61" y2="25" />
  <line x1="40" y1="30" x2="65" y2="30" />
  <line x1="43" y1="33" x2="68" y2="33" />
  <line x1="46" y1="36" x2="72" y2="36" />
  <line x1="50" y1="40" x2="75" y2="40" />
  <line x1="54" y1="44" x2="80" y2="44" />
  <line x1="58" y1="48" x2="83" y2="48" />
  <line x1="61" y1="51" x2="87" y2="51" />
  <line x1="65" y1="55" x2="90" y2="55" />
  <line x1="69" y1="59" x2="94" y2="59" />
  <line x1="72" y1="62" x2="98" y2="62" />
  <line x1="76" y1="66" x2="101" y2="66" />
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="35" y1="0" x2="106" y2="70" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.0,0.0 35.41261651458248,0.0 106.00085180870013,70.58823529411765 80.58823529411765,70.58823529411765" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />
  <line x1="80" y1="71" x2="106" y2="71" />
  <line x1="80" y1="72" x2="106" y2="72" />
  <line x1="80" y1="74" x2="106" y2="74" />
  <line x1="80" y1="75" x2="106" y2="75" />
  <line x1="80" y1="77" x2="106" y2="77" />
  <line x1="80" y1="78" x2="106" y2="78" />
  <line x1="80" y1="79" x2="106" y2="79" />
  <line x1="80" y1="80" x2="106" y2="80" />
  <line x1="80" y1="81" x2="106" y2="81" />
  <line x1="80" y1="82" x2="106" y2="82" />
  <line x1="80" y1="84" x2="106" y2="84" />
  <line x1="80" y1="85" x2="106" y2="85" />
  <line x1="80" y1="86" x2="106" y2="86" />
  <line x1="80" y1="88" x2="106" y2="88" />
  <line x1="80" y1="89" x2="106" y2="89" />
  <line x1="80" y1="90" x2="106" y2="90" />
  <line x1="80" y1="92" x2="106" y2="92" />
  <line x1="80" y1="94" x2="106" y2="94" />
  <line x1="80" y1="96" x2="106" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />
  <line x1="106" y1="70" x2="106" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="80.58823529411765,70.58823529411765 106.00085180870013,70.58823529411765 106.00085180870013,96.00085180870013 80.58823529411765,96.00085180870013" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Text -->
  <text x="93.294544" y="116.000852" font-size="1.0rem" font-weight="100" text-anchor="middle" >2</text>
  <text x="126.000852" y="83.294544" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,126.000852,83.294544)">2784</text>
  <text x="35.294118" y="80.706734" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(45,35.294118,80.706734)">52226568</text>
</svg>
</td>
</tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>call_GQ</span></div><div class='xr-var-dims'>(variants, samples)</div><div class='xr-var-dtype'>int16</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(600000, 81), meta=np.ndarray&gt;</div><input id='attrs-2a736ec7-6bd3-462f-894f-1db2461e98e4' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-2a736ec7-6bd3-462f-894f-1db2461e98e4' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d788f2fe-54fe-4266-b19d-dc86fa9b0ef6' class='xr-var-data-in' type='checkbox'><label for='data-d788f2fe-54fe-4266-b19d-dc86fa9b0ef6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 290.80 GB </td> <td> 122.40 MB </td></tr>
    <tr><th> Shape </th><td> (52226568, 2784) </td> <td> (600000, 102) </td></tr>
    <tr><th> Count </th><td> 16774 Tasks </td><td> 5032 Chunks </td></tr>
    <tr><th> Type </th><td> int16 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="75" height="170" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="25" y2="0" style="stroke-width:2" />
  <line x1="0" y1="5" x2="25" y2="5" />
  <line x1="0" y1="11" x2="25" y2="11" />
  <line x1="0" y1="17" x2="25" y2="17" />
  <line x1="0" y1="24" x2="25" y2="24" />
  <line x1="0" y1="31" x2="25" y2="31" />
  <line x1="0" y1="37" x2="25" y2="37" />
  <line x1="0" y1="44" x2="25" y2="44" />
  <line x1="0" y1="51" x2="25" y2="51" />
  <line x1="0" y1="56" x2="25" y2="56" />
  <line x1="0" y1="62" x2="25" y2="62" />
  <line x1="0" y1="68" x2="25" y2="68" />
  <line x1="0" y1="75" x2="25" y2="75" />
  <line x1="0" y1="82" x2="25" y2="82" />
  <line x1="0" y1="88" x2="25" y2="88" />
  <line x1="0" y1="93" x2="25" y2="93" />
  <line x1="0" y1="100" x2="25" y2="100" />
  <line x1="0" y1="106" x2="25" y2="106" />
  <line x1="0" y1="113" x2="25" y2="113" />
  <line x1="0" y1="120" x2="25" y2="120" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="120" style="stroke-width:2" />
  <line x1="0" y1="0" x2="0" y2="120" />
  <line x1="2" y1="0" x2="2" y2="120" />
  <line x1="3" y1="0" x2="3" y2="120" />
  <line x1="4" y1="0" x2="4" y2="120" />
  <line x1="6" y1="0" x2="6" y2="120" />
  <line x1="8" y1="0" x2="8" y2="120" />
  <line x1="9" y1="0" x2="9" y2="120" />
  <line x1="9" y1="0" x2="9" y2="120" />
  <line x1="11" y1="0" x2="11" y2="120" />
  <line x1="12" y1="0" x2="12" y2="120" />
  <line x1="13" y1="0" x2="13" y2="120" />
  <line x1="15" y1="0" x2="15" y2="120" />
  <line x1="16" y1="0" x2="16" y2="120" />
  <line x1="17" y1="0" x2="17" y2="120" />
  <line x1="18" y1="0" x2="18" y2="120" />
  <line x1="20" y1="0" x2="20" y2="120" />
  <line x1="21" y1="0" x2="21" y2="120" />
  <line x1="23" y1="0" x2="23" y2="120" />
  <line x1="25" y1="0" x2="25" y2="120" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 25.41261651458248,0.0 25.41261651458248,120.0 0.0,120.0" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Text -->
  <text x="12.706308" y="140.000000" font-size="1.0rem" font-weight="100" text-anchor="middle" >2784</text>
  <text x="45.412617" y="60.000000" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,45.412617,60.000000)">52226568</text>
</svg>
</td>
</tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>call_MQ</span></div><div class='xr-var-dims'>(variants, samples)</div><div class='xr-var-dtype'>int16</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(600000, 81), meta=np.ndarray&gt;</div><input id='attrs-cf77e2c4-e167-4732-817a-a0d6231c8f72' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-cf77e2c4-e167-4732-817a-a0d6231c8f72' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-068b5e6d-7c84-4938-922e-5df5b730d905' class='xr-var-data-in' type='checkbox'><label for='data-068b5e6d-7c84-4938-922e-5df5b730d905' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 290.80 GB </td> <td> 122.40 MB </td></tr>
    <tr><th> Shape </th><td> (52226568, 2784) </td> <td> (600000, 102) </td></tr>
    <tr><th> Count </th><td> 16774 Tasks </td><td> 5032 Chunks </td></tr>
    <tr><th> Type </th><td> int16 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="75" height="170" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="25" y2="0" style="stroke-width:2" />
  <line x1="0" y1="5" x2="25" y2="5" />
  <line x1="0" y1="11" x2="25" y2="11" />
  <line x1="0" y1="17" x2="25" y2="17" />
  <line x1="0" y1="24" x2="25" y2="24" />
  <line x1="0" y1="31" x2="25" y2="31" />
  <line x1="0" y1="37" x2="25" y2="37" />
  <line x1="0" y1="44" x2="25" y2="44" />
  <line x1="0" y1="51" x2="25" y2="51" />
  <line x1="0" y1="56" x2="25" y2="56" />
  <line x1="0" y1="62" x2="25" y2="62" />
  <line x1="0" y1="68" x2="25" y2="68" />
  <line x1="0" y1="75" x2="25" y2="75" />
  <line x1="0" y1="82" x2="25" y2="82" />
  <line x1="0" y1="88" x2="25" y2="88" />
  <line x1="0" y1="93" x2="25" y2="93" />
  <line x1="0" y1="100" x2="25" y2="100" />
  <line x1="0" y1="106" x2="25" y2="106" />
  <line x1="0" y1="113" x2="25" y2="113" />
  <line x1="0" y1="120" x2="25" y2="120" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="120" style="stroke-width:2" />
  <line x1="0" y1="0" x2="0" y2="120" />
  <line x1="2" y1="0" x2="2" y2="120" />
  <line x1="3" y1="0" x2="3" y2="120" />
  <line x1="4" y1="0" x2="4" y2="120" />
  <line x1="6" y1="0" x2="6" y2="120" />
  <line x1="8" y1="0" x2="8" y2="120" />
  <line x1="9" y1="0" x2="9" y2="120" />
  <line x1="9" y1="0" x2="9" y2="120" />
  <line x1="11" y1="0" x2="11" y2="120" />
  <line x1="12" y1="0" x2="12" y2="120" />
  <line x1="13" y1="0" x2="13" y2="120" />
  <line x1="15" y1="0" x2="15" y2="120" />
  <line x1="16" y1="0" x2="16" y2="120" />
  <line x1="17" y1="0" x2="17" y2="120" />
  <line x1="18" y1="0" x2="18" y2="120" />
  <line x1="20" y1="0" x2="20" y2="120" />
  <line x1="21" y1="0" x2="21" y2="120" />
  <line x1="23" y1="0" x2="23" y2="120" />
  <line x1="25" y1="0" x2="25" y2="120" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 25.41261651458248,0.0 25.41261651458248,120.0 0.0,120.0" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Text -->
  <text x="12.706308" y="140.000000" font-size="1.0rem" font-weight="100" text-anchor="middle" >2784</text>
  <text x="45.412617" y="60.000000" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,45.412617,60.000000)">52226568</text>
</svg>
</td>
</tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>call_AD</span></div><div class='xr-var-dims'>(variants, samples, alleles)</div><div class='xr-var-dtype'>int16</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(300000, 50, 4), meta=np.ndarray&gt;</div><input id='attrs-823384d8-e437-417d-b1c7-8e7c9cc0e6c2' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-823384d8-e437-417d-b1c7-8e7c9cc0e6c2' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7f31a023-e0ad-41f8-a730-11f3349c791f' class='xr-var-data-in' type='checkbox'><label for='data-7f31a023-e0ad-41f8-a730-11f3349c791f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 1.16 TB </td> <td> 120.00 MB </td></tr>
    <tr><th> Shape </th><td> (52226568, 2784, 4) </td> <td> (300000, 50, 4) </td></tr>
    <tr><th> Count </th><td> 24491 Tasks </td><td> 11900 Chunks </td></tr>
    <tr><th> Type </th><td> int16 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="156" height="146" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="10" y1="1" x2="80" y2="71" />
  <line x1="10" y1="2" x2="80" y2="73" />
  <line x1="10" y1="3" x2="80" y2="74" />
  <line x1="10" y1="4" x2="80" y2="75" />
  <line x1="10" y1="5" x2="80" y2="76" />
  <line x1="10" y1="7" x2="80" y2="78" />
  <line x1="10" y1="9" x2="80" y2="79" />
  <line x1="10" y1="10" x2="80" y2="80" />
  <line x1="10" y1="11" x2="80" y2="82" />
  <line x1="10" y1="12" x2="80" y2="83" />
  <line x1="10" y1="14" x2="80" y2="84" />
  <line x1="10" y1="15" x2="80" y2="86" />
  <line x1="10" y1="16" x2="80" y2="87" />
  <line x1="10" y1="18" x2="80" y2="88" />
  <line x1="10" y1="18" x2="80" y2="89" />
  <line x1="10" y1="20" x2="80" y2="91" />
  <line x1="10" y1="21" x2="80" y2="92" />
  <line x1="10" y1="23" x2="80" y2="94" />
  <line x1="10" y1="25" x2="80" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="10" y2="25" style="stroke-width:2" />
  <line x1="13" y1="3" x2="13" y2="29" />
  <line x1="17" y1="7" x2="17" y2="32" />
  <line x1="20" y1="10" x2="20" y2="36" />
  <line x1="24" y1="14" x2="24" y2="40" />
  <line x1="28" y1="18" x2="28" y2="44" />
  <line x1="32" y1="22" x2="32" y2="47" />
  <line x1="35" y1="25" x2="35" y2="51" />
  <line x1="39" y1="29" x2="39" y2="55" />
  <line x1="43" y1="33" x2="43" y2="58" />
  <line x1="47" y1="37" x2="47" y2="62" />
  <line x1="50" y1="40" x2="50" y2="66" />
  <line x1="54" y1="44" x2="54" y2="70" />
  <line x1="58" y1="48" x2="58" y2="73" />
  <line x1="61" y1="51" x2="61" y2="77" />
  <line x1="65" y1="55" x2="65" y2="81" />
  <line x1="69" y1="59" x2="69" y2="85" />
  <line x1="73" y1="63" x2="73" y2="88" />
  <line x1="76" y1="66" x2="76" y2="92" />
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.0,0.0 80.58823529411765,70.58823529411765 80.58823529411765,96.00085180870013 10.0,25.41261651458248" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="35" y2="0" style="stroke-width:2" />
  <line x1="13" y1="3" x2="39" y2="3" />
  <line x1="17" y1="7" x2="42" y2="7" />
  <line x1="20" y1="10" x2="46" y2="10" />
  <line x1="24" y1="14" x2="50" y2="14" />
  <line x1="28" y1="18" x2="54" y2="18" />
  <line x1="32" y1="22" x2="57" y2="22" />
  <line x1="35" y1="25" x2="61" y2="25" />
  <line x1="39" y1="29" x2="65" y2="29" />
  <line x1="43" y1="33" x2="68" y2="33" />
  <line x1="47" y1="37" x2="72" y2="37" />
  <line x1="50" y1="40" x2="76" y2="40" />
  <line x1="54" y1="44" x2="80" y2="44" />
  <line x1="58" y1="48" x2="83" y2="48" />
  <line x1="61" y1="51" x2="87" y2="51" />
  <line x1="65" y1="55" x2="91" y2="55" />
  <line x1="69" y1="59" x2="95" y2="59" />
  <line x1="73" y1="63" x2="98" y2="63" />
  <line x1="76" y1="66" x2="102" y2="66" />
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="35" y1="0" x2="106" y2="70" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.0,0.0 35.41261651458248,0.0 106.00085180870013,70.58823529411765 80.58823529411765,70.58823529411765" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />
  <line x1="80" y1="71" x2="106" y2="71" />
  <line x1="80" y1="73" x2="106" y2="73" />
  <line x1="80" y1="74" x2="106" y2="74" />
  <line x1="80" y1="75" x2="106" y2="75" />
  <line x1="80" y1="76" x2="106" y2="76" />
  <line x1="80" y1="78" x2="106" y2="78" />
  <line x1="80" y1="79" x2="106" y2="79" />
  <line x1="80" y1="80" x2="106" y2="80" />
  <line x1="80" y1="82" x2="106" y2="82" />
  <line x1="80" y1="83" x2="106" y2="83" />
  <line x1="80" y1="84" x2="106" y2="84" />
  <line x1="80" y1="86" x2="106" y2="86" />
  <line x1="80" y1="87" x2="106" y2="87" />
  <line x1="80" y1="88" x2="106" y2="88" />
  <line x1="80" y1="89" x2="106" y2="89" />
  <line x1="80" y1="91" x2="106" y2="91" />
  <line x1="80" y1="92" x2="106" y2="92" />
  <line x1="80" y1="94" x2="106" y2="94" />
  <line x1="80" y1="96" x2="106" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />
  <line x1="106" y1="70" x2="106" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="80.58823529411765,70.58823529411765 106.00085180870013,70.58823529411765 106.00085180870013,96.00085180870013 80.58823529411765,96.00085180870013" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Text -->
  <text x="93.294544" y="116.000852" font-size="1.0rem" font-weight="100" text-anchor="middle" >4</text>
  <text x="126.000852" y="83.294544" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,126.000852,83.294544)">2784</text>
  <text x="35.294118" y="80.706734" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(45,35.294118,80.706734)">52226568</text>
</svg>
</td>
</tr>
</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>call_genotype_mask</span></div><div class='xr-var-dims'>(variants, samples, ploidy)</div><div class='xr-var-dtype'>bool</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(600000, 81, 2), meta=np.ndarray&gt;</div><input id='attrs-d88a79f5-05d8-4e1f-b594-3602a63665ce' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-d88a79f5-05d8-4e1f-b594-3602a63665ce' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-2f5152b1-1ef2-41d1-927a-9386857fb291' class='xr-var-data-in' type='checkbox'><label for='data-2f5152b1-1ef2-41d1-927a-9386857fb291' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 290.80 GB </td> <td> 122.40 MB </td></tr>
    <tr><th> Shape </th><td> (52226568, 2784, 2) </td> <td> (600000, 102, 2) </td></tr>
    <tr><th> Count </th><td> 21806 Tasks </td><td> 5032 Chunks </td></tr>
    <tr><th> Type </th><td> bool </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="156" height="146" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="10" y1="0" x2="80" y2="71" />
  <line x1="10" y1="2" x2="80" y2="72" />
  <line x1="10" y1="3" x2="80" y2="74" />
  <line x1="10" y1="4" x2="80" y2="75" />
  <line x1="10" y1="6" x2="80" y2="77" />
  <line x1="10" y1="8" x2="80" y2="78" />
  <line x1="10" y1="9" x2="80" y2="79" />
  <line x1="10" y1="9" x2="80" y2="80" />
  <line x1="10" y1="11" x2="80" y2="81" />
  <line x1="10" y1="12" x2="80" y2="82" />
  <line x1="10" y1="13" x2="80" y2="84" />
  <line x1="10" y1="15" x2="80" y2="85" />
  <line x1="10" y1="16" x2="80" y2="86" />
  <line x1="10" y1="17" x2="80" y2="88" />
  <line x1="10" y1="18" x2="80" y2="89" />
  <line x1="10" y1="20" x2="80" y2="90" />
  <line x1="10" y1="21" x2="80" y2="92" />
  <line x1="10" y1="23" x2="80" y2="94" />
  <line x1="10" y1="25" x2="80" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="10" y2="25" style="stroke-width:2" />
  <line x1="13" y1="3" x2="13" y2="28" />
  <line x1="16" y1="6" x2="16" y2="32" />
  <line x1="20" y1="10" x2="20" y2="35" />
  <line x1="24" y1="14" x2="24" y2="39" />
  <line x1="28" y1="18" x2="28" y2="43" />
  <line x1="31" y1="21" x2="31" y2="47" />
  <line x1="35" y1="25" x2="35" y2="51" />
  <line x1="40" y1="30" x2="40" y2="55" />
  <line x1="43" y1="33" x2="43" y2="58" />
  <line x1="46" y1="36" x2="46" y2="62" />
  <line x1="50" y1="40" x2="50" y2="65" />
  <line x1="54" y1="44" x2="54" y2="70" />
  <line x1="58" y1="48" x2="58" y2="73" />
  <line x1="61" y1="51" x2="61" y2="77" />
  <line x1="65" y1="55" x2="65" y2="80" />
  <line x1="69" y1="59" x2="69" y2="84" />
  <line x1="72" y1="62" x2="72" y2="88" />
  <line x1="76" y1="66" x2="76" y2="91" />
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.0,0.0 80.58823529411765,70.58823529411765 80.58823529411765,96.00085180870013 10.0,25.41261651458248" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="10" y1="0" x2="35" y2="0" style="stroke-width:2" />
  <line x1="13" y1="3" x2="38" y2="3" />
  <line x1="16" y1="6" x2="42" y2="6" />
  <line x1="20" y1="10" x2="45" y2="10" />
  <line x1="24" y1="14" x2="49" y2="14" />
  <line x1="28" y1="18" x2="53" y2="18" />
  <line x1="31" y1="21" x2="57" y2="21" />
  <line x1="35" y1="25" x2="61" y2="25" />
  <line x1="40" y1="30" x2="65" y2="30" />
  <line x1="43" y1="33" x2="68" y2="33" />
  <line x1="46" y1="36" x2="72" y2="36" />
  <line x1="50" y1="40" x2="75" y2="40" />
  <line x1="54" y1="44" x2="80" y2="44" />
  <line x1="58" y1="48" x2="83" y2="48" />
  <line x1="61" y1="51" x2="87" y2="51" />
  <line x1="65" y1="55" x2="90" y2="55" />
  <line x1="69" y1="59" x2="94" y2="59" />
  <line x1="72" y1="62" x2="98" y2="62" />
  <line x1="76" y1="66" x2="101" y2="66" />
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="10" y1="0" x2="80" y2="70" style="stroke-width:2" />
  <line x1="35" y1="0" x2="106" y2="70" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="10.0,0.0 35.41261651458248,0.0 106.00085180870013,70.58823529411765 80.58823529411765,70.58823529411765" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Horizontal lines -->
  <line x1="80" y1="70" x2="106" y2="70" style="stroke-width:2" />
  <line x1="80" y1="71" x2="106" y2="71" />
  <line x1="80" y1="72" x2="106" y2="72" />
  <line x1="80" y1="74" x2="106" y2="74" />
  <line x1="80" y1="75" x2="106" y2="75" />
  <line x1="80" y1="77" x2="106" y2="77" />
  <line x1="80" y1="78" x2="106" y2="78" />
  <line x1="80" y1="79" x2="106" y2="79" />
  <line x1="80" y1="80" x2="106" y2="80" />
  <line x1="80" y1="81" x2="106" y2="81" />
  <line x1="80" y1="82" x2="106" y2="82" />
  <line x1="80" y1="84" x2="106" y2="84" />
  <line x1="80" y1="85" x2="106" y2="85" />
  <line x1="80" y1="86" x2="106" y2="86" />
  <line x1="80" y1="88" x2="106" y2="88" />
  <line x1="80" y1="89" x2="106" y2="89" />
  <line x1="80" y1="90" x2="106" y2="90" />
  <line x1="80" y1="92" x2="106" y2="92" />
  <line x1="80" y1="94" x2="106" y2="94" />
  <line x1="80" y1="96" x2="106" y2="96" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="80" y1="70" x2="80" y2="96" style="stroke-width:2" />
  <line x1="106" y1="70" x2="106" y2="96" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="80.58823529411765,70.58823529411765 106.00085180870013,70.58823529411765 106.00085180870013,96.00085180870013 80.58823529411765,96.00085180870013" style="fill:#8B4903A0;stroke-width:0"/>

  <!-- Text -->
  <text x="93.294544" y="116.000852" font-size="1.0rem" font-weight="100" text-anchor="middle" >2</text>
  <text x="126.000852" y="83.294544" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,126.000852,83.294544)">2784</text>
  <text x="35.294118" y="80.706734" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(45,35.294118,80.706734)">52226568</text>
</svg>
</td>
</tr>
</table></div></li></ul></div></li><li class='xr-section-item'><input id='section-9838f9f4-97c5-4427-84a9-4b8382fa499e' class='xr-section-summary-in' type='checkbox'  checked><label for='section-9838f9f4-97c5-4427-84a9-4b8382fa499e' class='xr-section-summary' >Attributes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'><dt><span>contigs :</span></dt><dd>(&#x27;2R&#x27;, &#x27;2L&#x27;, &#x27;3R&#x27;, &#x27;3L&#x27;, &#x27;X&#x27;)</dd></dl></div></li></ul></div></div></div></div>
</div>
<p>The arrays within this dataset can be accessed as shown below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pos</span> <span class="o">=</span> <span class="n">ds_snps</span><span class="p">[</span><span class="s1">&#39;variant_position&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">data</span>
<span class="n">pos</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table>
<tr>
<td>
<table>
  <thead>
    <tr><td> </td><th> Array </th><th> Chunk </th></tr>
  </thead>
  <tbody>
    <tr><th> Bytes </th><td> 208.91 MB </td> <td> 134.22 MB </td></tr>
    <tr><th> Shape </th><td> (52226568,) </td> <td> (33554432,) </td></tr>
    <tr><th> Count </th><td> 2 Tasks </td><td> 2 Chunks </td></tr>
    <tr><th> Type </th><td> int32 </td><td> numpy.ndarray </td></tr>
  </tbody>
</table>
</td>
<td>
<svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="77" y1="0" x2="77" y2="25" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >52226568</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(0,140.000000,12.706308)">1</text>
</svg>
</td>
</tr>
</table></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gt</span> <span class="o">=</span> <span class="n">allel</span><span class="o">.</span><span class="n">GenotypeDaskArray</span><span class="p">(</span><span class="n">ds_snps</span><span class="p">[</span><span class="s1">&#39;call_genotype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
<span class="n">gt</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="allel allel-DisplayAs2D"><span>&lt;GenotypeDaskArray shape=(52226568, 2784, 2) dtype=int8&gt;</span><table><thead><tr><th></th><th style="text-align: center">0</th><th style="text-align: center">1</th><th style="text-align: center">2</th><th style="text-align: center">3</th><th style="text-align: center">4</th><th style="text-align: center">...</th><th style="text-align: center">2779</th><th style="text-align: center">2780</th><th style="text-align: center">2781</th><th style="text-align: center">2782</th><th style="text-align: center">2783</th></tr></thead><tbody><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">0</th><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">...</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">1</th><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">...</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">2</th><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">...</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">...</th><td style="text-align: center" colspan="12">...</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">52226565</th><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">0/0</td><td style="text-align: center">...</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">0/0</td><td style="text-align: center">./.</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">52226566</th><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">0/0</td><td style="text-align: center">...</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">0/0</td><td style="text-align: center">./.</td><td style="text-align: center">0/0</td></tr><tr><th style="text-align: center; background-color: white; border-right: 1px solid black; ">52226567</th><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">0/0</td><td style="text-align: center">...</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td><td style="text-align: center">0/0</td><td style="text-align: center">./.</td><td style="text-align: center">./.</td></tr></tbody></table></div></div></div>
</div>
</div>
<div class="section" id="example-computation">
<h2>Example computation<a class="headerlink" href="#example-computation" title="Permalink to this headline">¶</a></h2>
<p>Here’s an example computation to count the number of segregating SNPs on chromosome arm 3R that also pass gamb_colu_arab site filters. This may take a minute or two, because it is scanning genotype calls at millions of SNPs in hundreds of samples.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># choose chromosome arm</span>
<span class="n">contig</span> <span class="o">=</span> <span class="s2">&quot;3R&quot;</span>

<span class="c1"># choose site filter mask</span>
<span class="n">mask</span> <span class="o">=</span> <span class="s2">&quot;gamb_colu_arab&quot;</span>

<span class="c1"># choose sample sets</span>
<span class="n">sample_sets</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AG1000G-BF-A&quot;</span><span class="p">,</span> <span class="s2">&quot;AG1000G-BF-B&quot;</span><span class="p">]</span>

<span class="c1"># locate pass sites</span>
<span class="n">loc_pass</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">site_filters</span><span class="p">(</span><span class="n">contig</span><span class="o">=</span><span class="n">contig</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="n">mask</span><span class="p">)</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>

<span class="c1"># perform an allele count over genotypes</span>
<span class="n">gt</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">snp_genotypes</span><span class="p">(</span><span class="n">contig</span><span class="o">=</span><span class="n">contig</span><span class="p">,</span> <span class="n">sample_sets</span><span class="o">=</span><span class="n">sample_sets</span><span class="p">)</span>
<span class="n">gt</span> <span class="o">=</span> <span class="n">allel</span><span class="o">.</span><span class="n">GenotypeDaskArray</span><span class="p">(</span><span class="n">gt</span><span class="p">)</span>
<span class="n">ac</span> <span class="o">=</span> <span class="n">gt</span><span class="o">.</span><span class="n">count_alleles</span><span class="p">(</span><span class="n">max_allele</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

<span class="c1"># locate segregating sites</span>
<span class="n">loc_seg</span> <span class="o">=</span> <span class="n">ac</span><span class="o">.</span><span class="n">is_segregating</span><span class="p">()</span>

<span class="c1"># count segregating and pass sites</span>
<span class="n">n_pass_seg</span> <span class="o">=</span> <span class="n">da</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">loc_pass</span> <span class="o">&amp;</span> <span class="n">loc_seg</span><span class="p">)</span>

<span class="c1"># run the computation</span>
<span class="k">with</span> <span class="n">ProgressBar</span><span class="p">():</span>
    <span class="n">n_pass_seg</span> <span class="o">=</span> <span class="n">n_pass_seg</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>

<span class="n">n_pass_seg</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[########################################] | 100% Completed |  1min 41.2s
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>11226723
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="running-larger-computations">
<h2>Running larger computations<a class="headerlink" href="#running-larger-computations" title="Permalink to this headline">¶</a></h2>
<p>Please note that free cloud computing services such as Google Colab and MyBinder provide only limited computing resources. Thus although these services are able to efficiently read Ag3 data stored on Google Cloud, you may find that you run out of memory, or computations take a long time running on a single core. If you would like any suggestions regarding how to set up more powerful compute resources in the cloud, please feel free to get in touch via the <a class="reference external" href="https://github.com/malariagen/vector-data/discussions">malariagen/vector-data GitHub discussion board</a>.</p>
</div>
<div class="section" id="feedback-and-suggestions">
<h2>Feedback and suggestions<a class="headerlink" href="#feedback-and-suggestions" title="Permalink to this headline">¶</a></h2>
<p>If there are particular analyses you would like to run, or if you have other suggestions for useful documentation we could add to this site, we would love to know, please get in touch via the <a class="reference external" href="https://github.com/malariagen/vector-data/discussions">malariagen/vector-data GitHub discussion board</a>.</p>
</div>
<div class="section" id="api-docs">
<h2>API docs<a class="headerlink" href="#api-docs" title="Permalink to this headline">¶</a></h2>
<p>Here are the docstrings for the functions in the <code class="docutils literal notranslate"><span class="pre">malariagen_data</span></code> package that we used above.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">ag3</span><span class="o">.</span><span class="n">sample_sets</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on method sample_sets in module malariagen_data.ag3:

sample_sets(release=&#39;v3&#39;) method of malariagen_data.ag3.Ag3 instance
    Access the manifest of sample sets.
    
    Parameters
    ----------
    release : str
        Release identifier. Give &quot;v3&quot; to access the Ag1000G phase 3 data release.
    
    Returns
    -------
    df : pandas.DataFrame
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">ag3</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on method sample_metadata in module malariagen_data.ag3:

sample_metadata(sample_sets=&#39;v3_wild&#39;, species_calls=(&#39;20200422&#39;, &#39;aim&#39;)) method of malariagen_data.ag3.Ag3 instance
    Access sample metadata for one or more sample sets.
    
    Parameters
    ----------
    sample_sets : str or list of str
        Can be a sample set identifier (e.g., &quot;AG1000G-AO&quot;) or a list of sample set
        identifiers (e.g., [&quot;AG1000G-BF-A&quot;, &quot;AG1000G-BF-B&quot;]) or a release identifier (e.g.,
        &quot;v3&quot;) or a list of release identifiers.
    species_calls : (str, str), optional
        Include species calls in metadata.
    
    Returns
    -------
    df : pandas.DataFrame
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">ag3</span><span class="o">.</span><span class="n">snp_sites</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on method snp_sites in module malariagen_data.ag3:

snp_sites(contig, field=None, site_mask=None, site_filters=&#39;dt_20200416&#39;, inline_array=True, chunks=&#39;auto&#39;) method of malariagen_data.ag3.Ag3 instance
    Access SNP site data (positions and alleles).
    
    Parameters
    ----------
    contig : str
        Chromosome arm, e.g., &quot;3R&quot;.
    field : {&quot;POS&quot;, &quot;REF&quot;, &quot;ALT&quot;}, optional
        Array to access. If not provided, all three arrays POS, REF, ALT will be returned as a
        tuple.
    site_mask : {&quot;gamb_colu_arab&quot;, &quot;gamb_colu&quot;, &quot;arab&quot;}
        Site filters mask to apply.
    site_filters : str
        Site filters analysis version.
    inline_array : bool, optional
        Passed through to dask.array.from_array().
    chunks : str, optional
        If &#39;auto&#39; let dask decide chunk size. If &#39;native&#39; use native zarr chunks.
        Also can be a target size, e.g., &#39;200 MiB&#39;.
    
    Returns
    -------
    d : dask.array.Array or tuple of dask.array.Array
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">ag3</span><span class="o">.</span><span class="n">site_filters</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on method site_filters in module malariagen_data.ag3:

site_filters(contig, mask, field=&#39;filter_pass&#39;, analysis=&#39;dt_20200416&#39;, inline_array=True, chunks=&#39;auto&#39;) method of malariagen_data.ag3.Ag3 instance
    Access SNP site filters.
    
    Parameters
    ----------
    contig : str
        Chromosome arm, e.g., &quot;3R&quot;.
    mask : {&quot;gamb_colu_arab&quot;, &quot;gamb_colu&quot;, &quot;arab&quot;}
        Mask to use.
    field : str, optional
        Array to access.
    analysis : str, optional
        Site filters analysis version.
    inline_array : bool, optional
        Passed through to dask.from_array().
    chunks : str, optional
        If &#39;auto&#39; let dask decide chunk size. If &#39;native&#39; use native zarr chunks.
        Also can be a target size, e.g., &#39;200 MiB&#39;.
    
    Returns
    -------
    d : dask.array.Array
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">ag3</span><span class="o">.</span><span class="n">snp_genotypes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on method snp_genotypes in module malariagen_data.ag3:

snp_genotypes(contig, sample_sets=&#39;v3_wild&#39;, field=&#39;GT&#39;, site_mask=None, site_filters=&#39;dt_20200416&#39;, inline_array=True, chunks=&#39;auto&#39;) method of malariagen_data.ag3.Ag3 instance
    Access SNP genotypes and associated data.
    
    Parameters
    ----------
    contig : str
        Chromosome arm, e.g., &quot;3R&quot;.
    sample_sets : str or list of str
        Can be a sample set identifier (e.g., &quot;AG1000G-AO&quot;) or a list of sample set
        identifiers (e.g., [&quot;AG1000G-BF-A&quot;, &quot;AG1000G-BF-B&quot;]) or a release identifier (e.g.,
        &quot;v3&quot;) or a list of release identifiers.
    field : {&quot;GT&quot;, &quot;GQ&quot;, &quot;AD&quot;, &quot;MQ&quot;}
        Array to access.
    site_mask : {&quot;gamb_colu_arab&quot;, &quot;gamb_colu&quot;, &quot;arab&quot;}
        Site filters mask to apply.
    site_filters : str, optional
        Site filters analysis version.
    inline_array : bool, optional
        Passed through to dask.array.from_array().
    chunks : str, optional
        If &#39;auto&#39; let dask decide chunk size. If &#39;native&#39; use native zarr chunks.
        Also can be a target size, e.g., &#39;200 MiB&#39;.
    
    Returns
    -------
    d : dask.array.Array
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">ag3</span><span class="o">.</span><span class="n">snp_calls</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on method snp_calls in module malariagen_data.ag3:

snp_calls(contig, sample_sets=&#39;v3_wild&#39;, site_mask=None, site_filters=&#39;dt_20200416&#39;, inline_array=True, chunks=&#39;auto&#39;) method of malariagen_data.ag3.Ag3 instance
    Access SNP sites, site filters and genotype calls.
    
    Parameters
    ----------
    contig : str
        Chromosome arm, e.g., &quot;3R&quot;.
    sample_sets : str or list of str
        Can be a sample set identifier (e.g., &quot;AG1000G-AO&quot;) or a list of sample set
        identifiers (e.g., [&quot;AG1000G-BF-A&quot;, &quot;AG1000G-BF-B&quot;]) or a release identifier (e.g.,
        &quot;v3&quot;) or a list of release identifiers.
    site_mask : {&quot;gamb_colu_arab&quot;, &quot;gamb_colu&quot;, &quot;arab&quot;}
        Site filters mask to apply.
    site_filters : str
        Site filters analysis version.
    inline_array : bool, optional
        Passed through to dask.array.from_array().
    chunks : str, optional
        If &#39;auto&#39; let dask decide chunk size. If &#39;native&#39; use native zarr chunks.
        Also can be a target size, e.g., &#39;200 MiB&#39;.
    
    Returns
    -------
    ds : xarray.Dataset
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./ag3"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='left-prev' id="prev-link" href="intro.html" title="previous page">Introduction to Ag3</a>
    <a class='right-next' id="next-link" href="download.html" title="next page">Ag3 data downloads</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By MalariaGEN<br/>
        
            &copy; Copyright 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>